### PR TITLE
Feature / Add Juz / Siparah marks for Ayat. 

### PR DESCRIPTION
**Pull request by Dhul-Husni**
### Title of change
  - Add juz/siparah marks for ayaat. from request #652 and #371 
### Checklist

- [x] Unit tests written
- [x] Manually tested
- [x] Prettier & ESLint were run
- [x] New dependencies are included in `package-lock.json`

### Screenshot
screen shot 2018-03-26 at 12 04 57 am
screen shot 2018-03-26 at 12 03 38 am

### CODE REVIEW

**Comment by sabeurthabti** [comment]
@ahmedre @mmahalwy @ahmedelgabri can you guys have a look at this, and review the UI.thanks guys.

File: src/components/JuzMarker/index.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import styled from 'styled-components';
+import { renderToString } from 'react-dom/server';
```
Comment (mmahalwy) [inline src/components/JuzMarker/index.js]: This should not be needed as this can be rendered on the client

File: src/components/JuzMarker/index.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  width: 100%;
+  text-align: center;
+  display: block;
+`;
+
+const StyledAyah = styled.div`
+  margin-top: 3%;
+`;
+
+export const Decoration = ({ juzNumber }) => (
```
Comment (mmahalwy) [inline src/components/JuzMarker/index.js]: Can we make this a separate component in a separate file?

File: src/components/JuzMarker/markNumbers.js [hunk 3]
```diff
@@ -0,0 +1,155 @@
+export const juzStart = {
```
Comment (mmahalwy) [inline src/components/JuzMarker/markNumbers.js]: Where do these come from? Can you write a note about this?

File: src/components/JuzMarker/markNumbers.js [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  58: [1],
+  67: [1],
+  78: [1]
+};
+
+export const juzAndHizbArabicNum = [
+  /* Juz and hizb numbers in arabic ~zero - sixty~ */
+  '٠',
+  '١',
+
```
Comment (mmahalwy) [inline src/components/JuzMarker/markNumbers.js]: Empty lines needed?

File: package.json [hunk 5]
```diff
@@ -90,6 +94,7 @@
     "clean-webpack-plugin": "0.1.10",
     "compression": "1.6.2",
     "compression-webpack-plugin": "0.3.2",
+    "concat-map": "0.0.1",
```
Comment (mmahalwy) [inline package.json]: Why is this added?

File: package.json [hunk 6]
```diff
@@ -109,6 +114,7 @@
     "html-pdf": "^2.1.0",
     "html-webpack-plugin": "1.7.0",
     "http-proxy": "1.14.0",
+    "http-signature": "1.2.0",
```
Comment (mmahalwy) [inline package.json]: This too

File: package.json [hunk 7]
```diff
@@ -189,6 +195,7 @@
     "eslint-plugin-jsx-a11y": "2.2.3",
     "eslint-plugin-mocha": "4.8.0",
     "eslint-plugin-react": "6.8.0",
+    "identity-obj-proxy": "3.0.0",
```
Comment (mmahalwy) [inline package.json]: This too

File: src/components/JuzMarker/index.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        <b
+          {...bindTooltip}
+          id={`Juz-mark-${chapterId}-${verseNumber}`}
+          key={`Juz-${juzNumber}`}
+          className="col-xs-1 icon-juzMarker"
+          title={`${intl.formatMessage({
+            id: 'juz.index.heading',
+            defaultMessage: 'juz'
+          })} ${juzNumber}`}
+          // Used dangerously... to prevent onhover staying even after onmouseOut
```
Comment (ahmedelgabri) [inline src/components/JuzMarker/index.js]: This is not a valid JSX comment syntax & also JSX doesn't allow inline comments.

```jsx
{/* JSX Comment syntax */}
```

```js
// JavaScript single line comment

/* 
 JavaScript
 multi-line 
 comment
*/
```

and in your case, ESLint should show you a parsing error where you have a comment.

```jsx
<b
  {...bindTooltip}
  id={`Juz-mark-${chapterId}-${verseNumber}`}
  key={`Juz-${juzNumber}`}
  className="col-xs-1 icon-juzMarker"
  title={`${intl.formatMessage({
    id: 'juz.index.heading',
    defaultMessage: 'juz'
  })} ${juzNumber}`}
  {/* Used dangerously... to prevent onhover staying even after onmouseOut */}   dangerouslySetInnerHTML={{
    __html: renderToString(<Decoration juzNumber={juzNumber} />)
  }}
/>
```

File: src/components/JuzMarker/index.js (hunk 1 above)
Comment (Dhul-Husni) [reply src/components/JuzMarker/index.js]: @mmahalwy  dangerouslySetInnerHtml only takes in valid strings and not raw components
i need it to convert a certain component to a string and pass it down to dangerouslySetInnerHtml

File: src/components/JuzMarker/index.js (hunk 2 above)
Comment (Dhul-Husni) [reply src/components/JuzMarker/index.js]: On it

File: src/components/JuzMarker/markNumbers.js (hunk 3 above)
Comment (Dhul-Husni) [reply src/components/JuzMarker/markNumbers.js]: Okay i'll add a note and a link to the source

File: src/components/JuzMarker/markNumbers.js (hunk 4 above)
Comment (Dhul-Husni) [reply src/components/JuzMarker/markNumbers.js]: Not needed will remove them

File: package.json (hunk 5 above)
Comment (Dhul-Husni) [reply package.json]: Left over from previous try. i'll get rid of them

File: package.json (hunk 6 above)
Comment (Dhul-Husni) [reply package.json]: Same here

File: package.json (hunk 7 above)
Comment (Dhul-Husni) [reply package.json]: I need this for mocking svg file import in the tests

File: src/components/JuzMarker/index.js (hunk 8 above)
Comment (Dhul-Husni) [reply src/components/JuzMarker/index.js]: True @ahmedelgabri. i'll change it

File: src/components/JuzMarker/markNumbers.js (hunk 3 above)
Comment (naveed-ahmad) [reply src/components/JuzMarker/markNumbers.js]: Or we use api to get juz for any ayah, like we discussed on slack. https://quran.api-docs.io/v3/juz/get-all-juzs

File: src/components/JuzMarker/index.js (hunk 1 above)
Comment (Dhul-Husni) [reply src/components/JuzMarker/index.js]: passing the component directly returns an error. Something about it having to be a string

File: src/components/JuzMarker/index.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+const StyledAyah = styled.div`
+  margin-top: 3%;
+`;
+
+export const Decoration = ({ juzNumber }) => (
+  // Show all the decoration in one component
+  <div>
+    <StyledJuzArt src={juzMarkTop} alt="juz top Art" />
+    <StyledContainer>
+      <StyledJuzNumber>جزء</StyledJuzNumber>
```
Comment (naveed-ahmad) [inline src/components/JuzMarker/index.js]: localized the strings as well, in Urdu its سپارہ

Comment (Dhul-Husni) [reply src/components/JuzMarker/index.js]: @naveed-ahmad  internationalization for urdu is already there for on hover
screen shot 2018-03-26 at 11 33 01 am
Would love to keep it as is in the mushaf distributed by king fahad complex

File: package.json (hunk 7 above)
Comment (sabeurthabti) [reply package.json]: we should probably use `babel-plugin-transform-require-ignore`  we have an issue raised to remove some legacy isomorphic code.

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33010

File: src/components/JuzMarker/markNumbers.js (hunk 3 above)
Comment (Dhul-Husni) [reply src/components/JuzMarker/markNumbers.js]: @naveed-ahmad  I feel like its abit of an over kill calling an api for every ayat to check whether it 's a start of a juz
thoughts?

**Comment by Dhul-Husni** [comment]
rebuild

File: package.json (hunk 7 above)
Comment (mmahalwy) [reply package.json]: You can do it natively with jest, see: https://github.com/quran/quran.com-frontend/pull/962/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R42

Comment (Dhul-Husni) [reply package.json]: okay @mmahalwy will copy it into the PR

Comment (Dhul-Husni) [reply package.json]: Done!.

File: src/components/JuzMarker/decoration.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+import styled from 'styled-components';
+import PropTypes from 'prop-types';
+import { juzAndHizbArabicNum } from './markNumbers';
+
+const juzMarkTop = require('../../../static/images/juzTopArt.svg');
+const juzMarkBottom = require('../../../static/images/juzBottomArt.svg');
+
+const StyledContainer = styled.div`
+  padding: 0.05em;
```
Comment (mmahalwy) [inline src/components/JuzMarker/decoration.js]: `em` is hard to control cause it's context dependent. Can we change this to be a fixed px?

File: src/components/JuzMarker/decoration.js [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+  border: solid 1px #2CA4AB;
+  width: 0.659em;
```
Comment (mmahalwy) [inline src/components/JuzMarker/decoration.js]: This too

File: src/components/JuzMarker/decoration.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+  margin-left: calc((100% - 0.659em) / 2);
```
Comment (mmahalwy) [inline src/components/JuzMarker/decoration.js]: too

File: src/components/JuzMarker/decoration.js [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+  margin-top: 0.055em;
+  margin-bottom: 0.055em;
+}
+`;
+
+const StyledJuzArt = styled.img`
+  height: 30px;
+  width: 0.659em;
```
Comment (mmahalwy) [inline src/components/JuzMarker/decoration.js]: too

File: src/components/JuzMarker/decoration.js [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+const juzMarkTop = require('../../../static/images/juzTopArt.svg');
+const juzMarkBottom = require('../../../static/images/juzBottomArt.svg');
+
+const StyledContainer = styled.div`
+  padding: 0.05em;
+  border: solid 1px #2CA4AB;
+  width: 0.659em;
+  margin-left: calc((100% - 0.659em) / 2);
+  margin-top: 0.055em;
+  margin-bottom: 0.055em;
```
Comment (mmahalwy) [inline src/components/JuzMarker/decoration.js]: too ^^

File: src/components/JuzMarker/index.js (hunk 1 above)
Comment (mmahalwy) [reply src/components/JuzMarker/index.js]: Can you paste the error? I can help more that way. But I am pretty confident we should not be using this on the client

File: src/components/JuzMarker/index.js [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+          id={`Juz-mark-${chapterId}-${verseNumber}`}
+          key={`Juz-${juzNumber}`}
+          className="col-xs-1 icon-juzMarker"
+          title={`${intl.formatMessage({
+            id: 'juz.index.heading',
+            defaultMessage: 'juz'
+          })} ${juzNumber}`}
+          dangerouslySetInnerHTML={{
+            /* Used dangerously... to prevent onhover staying even after onmouseOut */
+            __html: renderToString(<Decoration juzNumber={juzNumber} />)
```
Comment (mmahalwy) [inline src/components/JuzMarker/index.js]: this too

File: src/components/JuzMarker/markNumbers.js (hunk 3 above)
Comment (mmahalwy) [reply src/components/JuzMarker/markNumbers.js]: Hmm... feels like this should be a bootstrap API call. Might be an overkill for this PR imho since this will have little cost to size for the time being until we make it into an API

**Comment by mmahalwy** [comment]
@Dhul-Husni this is so close! :) excited for it to ship

File: src/components/JuzMarker/index.js (hunk 1 above)
Comment (Dhul-Husni) [reply src/components/JuzMarker/index.js]: Asalamu Aleikum @mmahalwy 
screen shot 2018-03-27 at 11 41 56 am
The above is what i get when i remove the renderToString Function
     ``` dangerouslySetInnerHTML={{
        __html: <Decoration juzNumber={juzNumber} />
      }}```
I decided to do abit more research based on what you said earlier
<cite>But I am pretty confident we should not be using this on the client</cite>
and 've got a bit of an **exciting news**
The react docs says: Link (Here)[https://reactjs.org/docs/react-dom-server.html]
`renderToString()
ReactDOMServer.renderToString(element)
Render a React element to its initial HTML. React will return an HTML string. You can use this method to generate HTML on the server and send the markup down on the initial request for faster page loads and to allow search engines to crawl your pages for SEO purposes.`

File: src/components/JuzMarker/decoration.js (hunk 10 above)
Comment (Dhul-Husni) [reply src/components/JuzMarker/decoration.js]: i didn't know that
Thanks will work on it

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33011

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33013

File: src/components/JuzMarker/index.js (hunk 1 above)
Comment (mmahalwy) [reply src/components/JuzMarker/index.js]: I see why you're getting what you got in the screenshot, it's because you are turning a javascript class into a string, that's what `dangerouslySetInnerHTML` will do. Instead, `<Decoration juzNumber={juzNumber} />` should be a child of `<b />`. 

So:
```
<b
          {...bindTooltip}
          id={`Juz-mark-${chapterId}-${verseNumber}`}
          key={`Juz-${juzNumber}`}
          className="col-xs-1"
          title={`${intl.formatMessage({
            id: 'juz.index.heading',
            defaultMessage: 'juz'
          })} ${juzNumber}`}
        >
        <Decoration juzNumber={juzNumber} />
     </b>
```
should work

**Comment by Dhul-Husni** [comment]
@mmahalwy  will fix the tooltip bug in the next PR and make necessary changes inshaAllah

**Comment by naveed-ahmad** [comment]
rebuild

**Comment by ahmedre** [comment]
Deployed to: http://staging.quran.com:33021
